### PR TITLE
framework: fix wrapping resources unwrapping

### PR DESCRIPTION
**Pull request by kopiczko**

### CODE REVIEW

File: framework/resource/internal/wrapper_test.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	return nil
+}
+
+type testWrappingResource struct {
+	resource framework.Resource
+}
+
+func (r *testWrappingResource) Wrapped() framework.Resource {
+	return r.resource
+}
```
Comment (kopiczko) [inline framework/resource/internal/wrapper_test.go]: The only difference between the resources is this one has `Wrapped()` methods which fulfills `Wrapper` interface.

File: framework/resource/retryresource/resource.go [hunk 2]
```diff
@@ -38,13 +41,25 @@ func New(config Config) (*Resource, error) {
...
 
+	var name string
+	{
+		u, err := internal.Underlying(config.Resource)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		name = u.Name()
+	}
```
Comment (xh3b4sd) [inline framework/resource/retryresource/resource.go]: This here means the retry resource itself does not have any name. Why is that? Don't we use this name for anything? If not, I would like to call the internal `name` something like `wrappedName` or `underlyingName`, depending on the naming we want to apply now.

File: framework/resource/internal/wrapper.go [hunk 3]
```diff
@@ -1,7 +1,29 @@
...
+		} else {
+			return r, nil
+		}
+
+		// When more that 1000 interations, assume infifinite loop.
+		i++
+		if i > 1000 {
+			return nil, microerror.Maskf(loopDetectedError, "unwrapped 1000 times, assuming infite loop: resource = %s", r.Name())
+		}
+	}
```
Comment (xh3b4sd) [inline framework/resource/internal/wrapper.go]: Why does this have to be so complicated? What is wrong with the current implementation?

File: framework/crud_resource.go [hunk 4]
```diff
@@ -343,8 +343,3 @@ func (r *CRUDResource) EnsureDeleted(ctx context.Context, obj interface{}) error
 func (r *CRUDResource) Name() string {
 	return r.name
 }
-
-// TODO uncomment when Resource interface is redefined.
-//func (r *CRUDResource) Underlying() Resource {
-//	return r
-//}
```
Comment (xh3b4sd) [inline framework/crud_resource.go]: I did not follow that closely enough apparently. Does this mean we can remove the underlying methods from all of our resource implementations?

Comment (kopiczko) [reply framework/crud_resource.go]: Yes!

File: framework/resource/retryresource/resource.go (hunk 2 above)
Comment (kopiczko) [reply framework/resource/retryresource/resource.go]: It was never used. I told that already in my previous PR.
As you can see it was so confusing nobody noticed so far. `underlyingName` works for me.

File: framework/resource/internal/wrapper.go (hunk 3 above)
Comment (kopiczko) [reply framework/resource/internal/wrapper.go]: The current (simplified) implementation doesn't unwrap nested resources. I wouldn't call it complex. It just unwraps a resource wrapped by possibly many wrappers. Which in our case is retryresource and metricsresource (many :D), but it has to be ready for possible more wrappers.

When we have CRUDResourceOps and CRUDResource it is difficult to set Underling in the right place.

It is also difficult to explain to users why you need to add Underlying method. And why it should return itself. It took me time to realize that at first place. We're doing the workshop, going to write a blogpost. We have 65 stars (which is a lot in operatokits word) - I'd like to make the thing more user friendly, easier to digest and spread the word.

**Pull request merged**